### ISSUE TITLE
azure network vpn-gateway list The api-version '2016-03-30' is invalid

### ISSUE THREAD

**Issue by howie** [Azure/azure-xplat-cli#3015]
CLI Version: **0.10.2**
OS Type: Mac 
Installation via:  brew 
Mode: **ARM **
Environment: **AzureCloud**
Description:
`azure network vpn-gateway list` and  `azure network local-gateway list` return error with wrong api version.

Steps to reproduce:
1) Run `azure network vpn-gateway list`

Error stack trace:

error:   The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'.

2016-07-16T04:20:36.535Z:
{ Error: The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'.
  <<< async stack >>>
  at get__6 (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/commands/arm/network/localNetworkGateway._js:145:71)
  at show__4 (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/commands/arm/network/localNetworkGateway._js:115:24)
  at __106 (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/commands/arm/network/network._js:2222:20)
  <<< raw stack >>>
    at /usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/azure-arm-network/lib/operations/localNetworkGateways.js:473:19
    at retryCallback (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/ms-rest/lib/filters/exponentialRetryPolicyFilter.js:140:9)
    at handleRedirect (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/ms-rest/lib/filters/redirectFilter.js:36:9)
    at /usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/util/utils.js:563:7
    at handleRedirect (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/util/utils.js:540:9)
    at /usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/util/logging.js:339:7
    at Request._callback (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/ms-rest/lib/requestPipeline.js:125:16)
    at Request.self.callback (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/request/request.js:199:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)

  stack: [Getter/Setter],
  statusCode: 400,
  request:
   { rawResponse: false,
     queryString: {},
     method: 'GET',
     headers:
      { 'x-ms-client-request-id': 'a06e9a79-994e-463a-a914-336b1d6a3e34',
        'accept-language': 'en-US',
        'Content-Type': 'application/json; charset=utf-8',
        'user-agent': 'WindowsAzureXplatCLI/0.10.2' },
     url: 'https://management.azure.com/subscriptions/4b54ff9b-1b03-4a35-825d-43bf44ea3361/resourceGroups/providers/Microsoft.Network/localNetworkGateways/?api-version=2016-03-30',
     body: null },
  response:
   { body: '{"error":{"code":"InvalidApiVersionParameter","message":"The api-version \'2016-03-30\' is invalid. The supported versions are \'2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04\'."}}',
     headers:
      { 'cache-control': 'no-cache',
        pragma: 'no-cache',
        'content-type': 'application/json; charset=utf-8',
        expires: '-1',
        'x-ms-failure-cause': 'gateway',
        'x-ms-request-id': '86adb7ea-e87b-4764-bf6e-e0c319075f3f',
        'x-ms-correlation-request-id': '86adb7ea-e87b-4764-bf6e-e0c319075f3f',
        'x-ms-routing-request-id': 'EASTASIA:20160716T042036Z:86adb7ea-e87b-4764-bf6e-e0c319075f3f',
        'strict-transport-security': 'max-age=31536000; includeSubDomains',
        date: 'Sat, 16 Jul 2016 04:20:36 GMT',
        connection: 'close',
        'content-length': '256' },
     statusCode: 400 },
  code: 'InvalidApiVersionParameter',
  body:
   { code: 'InvalidApiVersionParameter',
     message: 'The api-version \'2016-03-30\' is invalid. The supported versions are \'2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04\'.' },
  __frame:
   { name: 'get__6',
     line: 140,
     file: '/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/commands/arm/network/localNetworkGateway._js',
     prev:
      { name: 'show__4',
        line: 113,
        file: '/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/commands/arm/network/localNetworkGateway._js',
        prev: [Object],
        calls: 1,
        active: false,
        offset: 2,
        col: 23 },
     calls: 0,
     active: false,
     offset: 5,
     col: 70 },
  rawStack: [Getter] }
Error: The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'.
  <<< async stack >>>
  at get__6 (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/commands/arm/network/localNetworkGateway._js:145:71)
  at show__4 (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/commands/arm/network/localNetworkGateway._js:115:24)
  at __106 (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/commands/arm/network/network._js:2222:20)
  <<< raw stack >>>
    at /usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/azure-arm-network/lib/operations/localNetworkGateways.js:473:19
    at retryCallback (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/ms-rest/lib/filters/exponentialRetryPolicyFilter.js:140:9)
    at handleRedirect (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/ms-rest/lib/filters/redirectFilter.js:36:9)
    at /usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/util/utils.js:563:7
    at handleRedirect (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/util/utils.js:540:9)
    at /usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/lib/util/logging.js:339:7
    at Request._callback (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/ms-rest/lib/requestPipeline.js:125:16)
    at Request.self.callback (/usr/local/Cellar/azure-cli/0.10.2/libexec/lib/node_modules/azure-cli/node_modules/request/request.js:199:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)

**Comment by amarzavery** [comment]
@howie - I am not able to repro. Both the commands work properly for me. Can you please uninstall the cli installed via brew and install via npm or the mac installer?

**Comment by howie** [comment]
Hi @amarzavery  
After reinstall by npm install -g azure-cli , still have this problem

howie@howiedeMBP ⮀ ~ ⮀ azure network vpn-gateway  list
info:    Executing command network vpn-gateway list
Resource group name:
+ Looking up virtual network gateways
error:   The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'.
error:   Error information has been recorded to /Users/howie/.azure/azure.err
error:   network vpn-gateway list command failed

By the way , Dose azure-cli have any plan to implement command to check gateway bandwidth usage ?

**Comment by ndelitski** [comment]
have exactly the same error when executing:
```
$ azure network nic delete --name my-nic-name
```

**Comment by amarzavery** [comment]
@MikhailTryakhov @haocs - can you guys take a look at this? I verified it on my mac by installing the latest version of azure-cli  (0.10.2) and the command works fine. 

@ndelitski @howie - Can you guys execute the command with -vv option and paste the content of execution over here. That might help us in our investigation.

**Comment by haocs** [comment]
@amarzavery ,  @MikhailTryakhov  - I got the same error on windows using version 0.10.2. `api-version 2016-03-03` seems not accepted by server. 
Please see the following `azure network vpn-gateway list` log with -vv option.
```javascript
D:\P\azure-xplat-cli>azure network vpn-gateway list -vv
info:    Executing command network vpn-gateway list
Resource group name:
verbose: Looking up virtual network gateways
silly:   requestOptions
silly:   {
silly:       rawResponse: false,
silly:       queryString: {},
silly:       method: 'GET',
silly:       headers: {
silly:           user-agent: 'WindowsAzureXplatCLI/0.10.2',
silly:           Content-Type: 'application/json; charset=utf-8',
silly:           accept-language: 'en-US',
silly:           x-ms-client-request-id: '7180842f-dc78-4839-80b8-3e06ad0b66cb'
silly:       },
silly:       url: 'https://management.azure.com/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/providers/Microsoft.Network/virtualNetworkGateways?api-version=2016-03-30',
silly:       body: null
silly:   }
silly:   returnObject
silly:   {
silly:       statusCode: 400,
silly:       header: {
silly:           cache-control: 'no-cache',
silly:           pragma: 'no-cache',
silly:           content-type: 'application/json; charset=utf-8',
silly:           expires: '-1',
silly:           x-ms-failure-cause: 'gateway',
silly:           x-ms-request-id: '1147c9a5-c5bd-4fba-bdb3-2d0c14a4437d',
silly:           x-ms-correlation-request-id: '1147c9a5-c5bd-4fba-bdb3-2d0c14a4437d',
silly:           x-ms-routing-request-id: 'CENTRALUS:20160718T214721Z:1147c9a5-c5bd-4fba-bdb3-2d0c14a4437d',
silly:           strict-transport-security: 'max-age=31536000; includeSubDomains',
silly:           date: 'Mon, 18 Jul 2016 21:47:21 GMT',
silly:           connection: 'close',
silly:           content-length: '256'
silly:       },
silly:       body: "{"error":{"code":"InvalidApiVersionParameter","message":"The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'."}}"
silly:   }
error:   The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'.
silly:   {
silly:       __frame: {
silly:           name: 'list__3',
silly:           line: 114,
silly:           file: 'C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\commands\\arm\\network\\virtualNetworkGateway.js',
silly:           prev: {
silly:               name: '__110',
silly:               line: 2318,
silly:               file: 'C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\commands\\arm\\network\\network.js',
silly:               prev: undefined,
silly:               calls: 1,
silly:               active: false,
silly:               offset: 5,
silly:               col: 18
silly:           },
silly:           calls: 0,
silly:           active: false,
silly:           offset: 6,
silly:           col: 69
silly:       },
silly:       response: {
silly:           body: "{"error":{"code":"InvalidApiVersionParameter","message":"The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'."}}",
silly:           headers: {
silly:               cache-control: 'no-cache',
silly:               pragma: 'no-cache',
silly:               content-type: 'application/json; charset=utf-8',
silly:               expires: '-1',
silly:               x-ms-failure-cause: 'gateway',
silly:               x-ms-request-id: '1147c9a5-c5bd-4fba-bdb3-2d0c14a4437d',
silly:               x-ms-correlation-request-id: '1147c9a5-c5bd-4fba-bdb3-2d0c14a4437d',
silly:               x-ms-routing-request-id: 'CENTRALUS:20160718T214721Z:1147c9a5-c5bd-4fba-bdb3-2d0c14a4437d',
silly:               strict-transport-security: 'max-age=31536000; includeSubDomains',
silly:               date: 'Mon, 18 Jul 2016 21:47:21 GMT',
silly:               connection: 'close',
silly:               content-length: '256'
silly:           },
silly:           statusCode: 400
silly:       },
silly:       message: "The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'.",
silly:       statusCode: 400,
silly:       body: { code: 'InvalidApiVersionParameter', message: "The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'." },
silly:       rawStack: "Error: The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'.\n    at C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\azure-arm-network\\lib\\operations\\virtualNetworkGateways.js:892:19\n    at retryCallback (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\ms-rest\\lib\\filters\\exponentialRetryPolicyFilter.js:140:9)\n    at handleRedirect (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\ms-rest\\lib\\filters\\redirectFilter.js:36:9)\n    at C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\util\\utils.js:563:7\n    at handleRedirect (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\util\\utils.js:540:9)\n    at C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\util\\logging.js:339:7\n    at Request._callback (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\ms-rest\\lib\\requestPipeline.js:125:16)\n    at Request.self.callback (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\request\\request.js:199:22)\n    at emitTwo (events.js:87:13)\n    at Request.emit (events.js:172:7)",
silly:       stack: "Error: The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'.\n  <<< async stack >>>\n  at list__3 (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\commands\\arm\\network\\virtualNetworkGateway.js:120:70)\n  at __110 (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\commands\\arm\\network\\network.js:2323:19)\n  <<< raw stack >>>\n    at C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\azure-arm-network\\lib\\operations\\virtualNetworkGateways.js:892:19\n    at retryCallback (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\ms-rest\\lib\\filters\\exponentialRetryPolicyFilter.js:140:9)\n    at handleRedirect (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\ms-rest\\lib\\filters\\redirectFilter.js:36:9)\n    at C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\util\\utils.js:563:7\n    at handleRedirect (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\util\\utils.js:540:9)\n    at C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\lib\\util\\logging.js:339:7\n    at Request._callback (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\ms-rest\\lib\\requestPipeline.js:125:16)\n    at Request.self.callback (C:\\Program Files (x86)\\Microsoft SDKs\\Azure\\CLI\\node_modules\\request\\request.js:199:22)\n    at emitTwo (events.js:87:13)\n    at Request.emit (events.js:172:7)",
silly:       request: {
silly:           rawResponse: false,
silly:           queryString: {},
silly:           method: 'GET',
silly:           headers: {
silly:               user-agent: 'WindowsAzureXplatCLI/0.10.2',
silly:               Content-Type: 'application/json; charset=utf-8',
silly:               accept-language: 'en-US',
silly:               x-ms-client-request-id: '7180842f-dc78-4839-80b8-3e06ad0b66cb'
silly:           },
silly:           url: 'https://management.azure.com/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/providers/Microsoft.Network/virtualNetworkGateways?api-version=2016-03-30',
silly:           body: null
silly:       },
silly:       code: 'InvalidApiVersionParameter'
silly:   }
error:   Error information has been recorded to C:\Users\hach\.azure\azure.err
verbose: Error: The api-version '2016-03-30' is invalid. The supported versions are '2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'.
  <<< async stack >>>
  at list__3 (C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\lib\commands\arm\network\virtualNetworkGateway.js:120:70)
  at __110 (C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\lib\commands\arm\network\network.js:2323:19)
  <<< raw stack >>>
    at C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\node_modules\azure-arm-network\lib\operations\virtualNetworkGateways.js:892:19
    at retryCallback (C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\node_modules\ms-rest\lib\filters\exponentialRetryPolicyFilter.js:140:9)
    at handleRedirect (C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\node_modules\ms-rest\lib\filters\redirectFilter.js:36:9)
    at C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\lib\util\utils.js:563:7
    at handleRedirect (C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\lib\util\utils.js:540:9)
    at C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\lib\util\logging.js:339:7
    at Request._callback (C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\node_modules\ms-rest\lib\requestPipeline.js:125:16)
    at Request.self.callback (C:\Program Files (x86)\Microsoft SDKs\Azure\CLI\node_modules\request\request.js:199:22)
    at emitTwo (events.js:87:13)
    at Request.emit (events.js:172:7)
error:   network vpn-gateway list command failed
```

**Comment by haocs** [comment]
@MikhailTryakhov - BTW.  usage info for command `vpn-gateway list` seems not accurate. 
`help:    Usage: network vpn-gateway list [options] <resource-group>` 
`resource-group` is marked as required however it is optional in execution.

**Comment by amarzavery** [comment]
@deepakrajendranmsft - Can you take a look at this? Why would api-version 2016-03-30 be invalid?

**Comment by amarzavery** [comment]
@deepakrajendranmsft @MikhailTryakhov - I found the problem. We should be throwing an error if the resourcegroupName is an empty string or string with spaces. Empty string is not null or undefined hence passes the sdk validation and creates an incorrect url. However, the message returned by the server about incorrect api-version is very misleading. 

The validation of required parameters (to be non empty string) should be fixed in the upcoming release for all the network commands in general.

**Comment by amarzavery** [comment]
https://github.com/Azure/azure-xplat-cli/issues/1847

**Comment by szliyang** [comment]
I met this issue as well. Eventually I able to overcome this issue via https://portal.azure.com

**Comment by amarzavery** [comment]
@ndelitski @howie @szliyang - The resouregroupName is a required parameter. Hence, when the CLI prompts you for one, please provide the correct resource group name and then press enter. Then the command should run fine. Hope this helps for now.

We have fixed the issue, by making the CLI throw an error if it receives a null or an empty string or a string with spaces as an input when it prompts for anything. This has been fixed in PR #3030 and it will be part of the August release.

**Comment by CristiamMorenoPosada** [comment]
Hi. I know this post is from last year, but i have a similar problem with the NodeJs SDK:
The api-version \'2016-03-30\' is invalid. The supported versions are \'2017-05-10,2017-05-01,2017-03-01,2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04\
